Allow user to specify colors in violin plots with constructor method

landoskape opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
This PR adds the option of specifying the fillcolor and linecolor of violinplots. It addresses
issue #27298 (and other discusssions and requests elsewhere).

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] Closes #27298 
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

oscargus commented:
Thanks!

Can you please add a test for this? I think either an image test, e.g. with a number of subplots using different color arguments, or an image comparison test where one check that using the arguments and setting the colors afterwards result in identical images.

oscargus commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8226,6 +8227,17 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
+        fillcolor : color or None; see :ref:`colors_def`
+          If provided, will set the fill color of the violin plots. The alpha
+          value of the fill is automatically set to 0.3.
+
```
Please see https://matplotlib.org/stable/devel/coding_guide.html#new-features-and-api-changes

Basically, `.. versionadded:: 3.9` should be added for these three arguments. However, see the link above to get the indentation correct (otherwise I would have added a suggestion, but most likely it will be wrong...)

landoskape replied:
Sure thing. Those lines have been added. I inherited the two space indent rather than the typical full tab for the descriptions because that is how the violinplot code was already structured. But all other details follow the example from the above link!

oscargus replied:
Sorry, I just meant that I couldn't provide a proper suggestion as it is more or less impossible to get the spacing correct in the GitHub comment.

These looks good!

oscargus commented:
I couldn't find a doc-page for that, but in https://github.com/matplotlib/matplotlib/tree/main/doc/users/next_whats_new there are small notes to be included in the next What's New. Please add one for this feature, ideally with an example showing how it works to highlight this new and useful feature.

landoskape commented:
> Thanks!
> 
> Can you please add a test for this? I think either an image test, e.g. with a number of subplots using different color arguments, or an image comparison test where one check that using the arguments and setting the colors afterwards result in identical images.
> 
> Edit: see https://matplotlib.org/stable/devel/testing.html#writing-an-image-comparison-test if more info on the testing is required.

Test added, I have never written one but I believe this should work as expected.

landoskape commented:
And now there's a whats new page. I think that covers your requests. Let me know if there's anything else to do! Thanks for your feedback and support.

landoskape commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8173,7 +8173,8 @@ def matshow(self, Z, **kwargs):
                    showmeans=False, showextrema=True, showmedians=False,
-                   quantiles=None, points=100, bw_method=None):
+                   quantiles=None, points=100, bw_method=None,
+                   fillcolor=None, linecolor=None, color=None):
```
Ah yes. I thought so but kept fill and line color to remain internally consistent with the previous violinplot variable names. Will update this to face and edge color.

landoskape commented:
I think that's everything @oscargus and @story645. Thanks for your help!

oscargus commented on doc/users/next_whats_new/violinplot_colors.rst:
```diff
@@ -0,0 +1,7 @@
+The ``~.Axes.violinplot`` constructor now accepts ``color``, ``facecolor``,
+and ``edgecolor`` as input arguments. This means that users can set the color
+of violinplots as they make them, rather than setting the color of individual
+objects aftewards.
```
```suggestion
objects afterwards.
```

oscargus commented on galleries/examples/statistics/customized_violin.py:
```diff
@@ -44,11 +44,10 @@ def set_axis_style(ax, labels):
 
 for pc in parts['bodies']:
-    pc.set_facecolor('#D43F3A')
     pc.set_edgecolor('black')
```
```suggestion
```
this should be removed?

oscargus reviewed (approved):
Some minor comments.

Maybe one should change both face- and edgecolor (to different values) in one of the subplots to confirm that it also works?

landoskape commented:
> Maybe one should change both face- and edgecolor (to different values) in one of the subplots to confirm that it also works?

Good point. I added two new subplot axes to the test - one that changes ``facecolor`` and ``edgecolor`` to different things, and one that checks if ``color`` overwrites the other two.

landoskape commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8332,6 +8353,23 @@ def violin(self, vpstats, positions=None, vert=True, widths=0.5,
         showmedians : bool, default: False
           If true, will toggle rendering of the medians.
 
+        facecolor : color or None; see :ref:`colors_def`
+          If provided, will set the face color of the violin plots. The alpha
+          value is automatically set to 0.3.
+
+          .. versionadded:: 3.9
+
+        edgecolor : color or None; see :ref:`colors_def`
+          If provided, will set the edge color of the violin plots (the
+          horizontal and vertical spines).
+
+          .. versionadded:: 3.9
+
+        color : color or None; see :ref:`colors_def`
+          If provided, will set (and overwrite) the facecolor and edgecolor.
+
+          .. versionadded:: 3.9
```
The `axes.violin()` method is not what the users calls to make violinplots -- they either call `pyplot.violinplot()` or `axes.violinplot()`, and both of these eventually call `axes.violin()`. So, the color arguments need to be in the `violinplot()` methods for it to be useful.

landoskape commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8226,6 +8227,23 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
+        color : color or None; see :ref:`colors_def`
+          If provided, will set (and overwrite) the facecolor and edgecolor.
+
+          .. versionadded:: 3.9
```
This is now the way it's organized

landoskape commented:
> I understand if you feel this is out of scope, but since you're already implementing these parameters how do you feel about implementing them vectorized? Meaning that you can also pass in a list of colors/facecolor/edgecolor like in bar?

Good idea! That's now implemented with documentation, a test for this feature, and an example for the gallery. It's implemented the same way as it is in the `bar` method, where a sequence of colors can be passed that has a different length than the data, but the violin method will repeat the colors.

ksunden commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8226,6 +8227,21 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
 
+        facecolor : color or list of colors or None; see :ref:`colors_def`
+          If provided, will set the face color(s) of the violin plots. The alpha
+          value is automatically set to 0.3.
```
I'm trying to figure out what the semantics _should_ be to remain consistent with current behavior... It feels wrong to override the alpha value of a user provided color...

Though I will also note that the `0.3` alpha is actually applied to the patch object as a whole, not its color values...

Makes me wonder if the "better" solution would be to just pass `**kwargs` to lower level artists (But I think that gets complicated quickly because some of them take the "edgecolor" as "color", for instance..., and other values would have to be filtered out likely...)

landoskape replied:
That's a good point. I think that if the user doesn't specify the alpha, it makes sense to set it to 0.3. That just looks good, and honestly I'd prefer that as someone who wants to use the violinplot method with these new features. 

I added a commit that changes the behavior: 
If the user specifies the alpha, then that is the alpha that is used. If the user doesn't specify the alpha, then it defaults to 0.3.

timhoffm reviewed (changes requested):
Sorry for being late to the party.

Up to now, styling was not configurable at all. That was simple. When we now start to add options, we should be extremely careful with the API. To be consistent, we need to look towards `bar()` as well as `boxplot()`.

#### `bar()` doc:
- `color` for the bar faces
- `edgecolor` for the bar edges
- `ecolors` for the errorbars
but in `kwargs`, we have the standard `color`, `edgecolor`, `facecolor` behavior, where `color` sets both edge and face. This conflicts with the color doc above. We need to check what's actually happening.

#### `boxplot()`
Styling is configured via dicts for the individual parts
- `boxprops`
- `capprops`
- `whiskerprops`
- `flierprops`
- `medianprops`
- `meanprops`
This allows very detailed configuration, but is cumbersome to use.

#### `violinplot()`
We have the following parts
- bodies
- cmeans
- cmins
- cmaxes
- cbars
- cmedians
- cquantiles

To follow the concept of `boxplot()`, we would need to make them available via `*props`, but that's quite a lift, and I'm unsure on the usefulness.

Generally, the proposed global color settings seem useful (and that doesn't mean we cannot add `*props` later or add color kwargs to `boxplot()`).

For now the two minimal requirements if have on API consistency are
- Check the `color` behavior of `bar()` and be consistent with it or argue why we should behave differently here.
- Is `edgecolor` the correct name? Does it color the `bodies` edges (if they have a finite width)? Should it?

landoskape commented:
@timhoffm I can tell you from my perspective as someone who thinks about matplotlib primarily as a user (this is my first PR) I'm very happy to have a plotting function that allows me to make something look good very quickly, but that has the option of more detailed customization if required. The way the violinplot works now (after this PR) seems to fit that criteria. You can make a violin or set of violins with a color scheme that works, in one line of code, then go and edit individual objects later if you need to. 

--

> For now the two minimal requirements if have on API consistency are
> 
> * Check the `color` behavior of `bar()` and be consistent with it or argue why we should behave differently here.

The color behavior between the new violinplot() and bar() are identical in terms of how the colors are handled within the function. However, bar() only has edgecolor and color as inputs. I changed the violinplot() to have edgecolor, facecolor, and color (which overwrites the other two) in response to @ksunden's point in issue #27298. This makes it consistent with patches.

In my opinion, this makes a lot of sense, because a user might want specific control over the lines and patch within the violin plot (i.e. using edgecolor/facecolor), or might want to make everything look the same (i.e. using color). That seems like a perfect compromise between customizability and simplicity. So if anything, maybe bar() should copy this behavior?

Extra point: I think this edgecolor/facecolor/color schema additionally makes sense given the way I plan to handle alpha values, which I'll explain in the response to @ksunden and @story645 in the above thread. 

--

> * Is `edgecolor` the correct name? Does it color the `bodies` edges (if they have a finite width)? Should it?

I changed the name to edgecolor from linecolor because @ksunden and @story645 pointed out that that's the typical name for the rest of the library. I agree with that reasoning.

As it is, the bodies by default do not have an edgeline for violinplots. This is how it was before and still is. I think this is great, because it creates a very clean looking plot, and if the user wants to, they can go in and add an edge to the bodies. We can add this as an example to the [customized violin plot](https://github.com/matplotlib/matplotlib/blob/main/galleries/examples/statistics/customized_violin.py) in the gallery for the interested user if you think that's a good idea.

ksunden commented:
I guess I had it in my head when I made the original suggestion that the main shape were by default drawn with edges, which isn't actually true.

I think Tim is right that `bar` is the closer analog here...

I do think that the most full solution would be to do something like `bodyprops=dict(...)`, `cmeanprops=dict(...)`, etc... (naming up to debate, as they are pluralized in the returns)... but I also agree that it is probably a bigger lift than we _really_ want.) That style is explicit and flexible (and something we do use in other parts of the library, particularly when there are multiple places `**kwargs` could be routed to, including `boxplot`, which is the closest analog to `violinplot`), but not necessarily the most friendly to use. 

`box` has `color` (defaults to next patch color), `edgecolor` (defaults to None, which translates to the same as facecolor, I think) as well as `ecolor` (defaults to black, used for error bars).

`color` is either a single color or a list of colors which is cycled.

landoskape commented:
Not to complicate things further but regarding the earlier discussion it also may be helpful to include an argument like `facealpha`. I'm happy to keep working on it (although I think I may tap out if the lift to ``bodyprops=dict(...)` is requested). But I think these decisions are above my pay grade so let me know what you all think.

Also, I'm not sure I understand why the tests failed. 
The python 3.9 on ubuntu-20.04 (with and without Minimum Versions) failed because: 
``FAILED lib/matplotlib/tests/test_backends_interactive.py::test_figure_leak_20490[time_mem0-{'MPLBACKEND': 'wxagg'}]``

And the Main Pytest Windows_py39 failed because:
``FAILED lib/matplotlib/tests/test_pickle.py::test_pickle_load_from_subprocess[png]``

But I'm not sure why that is happening or how it relates to the commits I've made.

oscargus commented:
> how it relates to the commits I've made.

It shouldn't be directly related. While it is weird that it seems to happen consistently on this PR (and some others), there is no obvious reason why it happens.

landoskape commented:
> > it also may be helpful to include an argument like facealpha.
> 
> We added support for `(color, alpha)` as a valid [color specification](https://matplotlib.org/devdocs/users/explain/colors/colors.html#sphx-glr-users-explain-colors-colors-py) precisely to support specifying alpha on specific components w/o growing kwargs #24691

Yep. I now removed the alpha handling so the alpha will just directly be controlled by `facecolor` without defaults to alternative values.

landoskape commented:
I just wanted to ping the reviewers in case there is continued interest in this PR. I'm happy to continue working on the changes to the violinplot constructor, but there was no clear agreement on how best to manage the additional color arguments. 

It seems to me like there is tension between which other plotting method ``violinplot`` should be consistent with -- should it be similar to ``barplot`` or ``boxplot`` (or something else)? And -- should the kwargs be expanded to include parameter control over each feature of the violinplot -- e.g. ``bodies``, ``cmeans``, ``cmins``, etc. or just have a single set of kwargs to control each feature of the plot. 

In my opinion as a user, I'm very happy to have a few simple kwargs (like in ``barplot``, i.e. just having ``color``, ``edgecolor``, ``facecolor``) and letting matplotlib apply them to each feature of the plot so it looks good with minimal effort. The output of the constructor still provides the option of using the feature dictionaries for further specification if desired by a user.

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8392,31 +8424,61 @@ def violin(self, vpstats, positions=None, vert=True, widths=0.5,
 
-        # Colors.
+        # Make a cycle of color to iterate through, using 'none' as fallback
+        def cycle_color(color):
```
I do not understand this function, `itertools.cycle` repeats no we never hit the fallback to `none`.

landoskape replied:
It is meant to copy the functionality in ``bar`` (see [here](https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/axes/_axes.py#L2500)), in which ``"none"`` is a possible input to ``mcolors.to_rgba_array()`` that returns ``[]``. This is the specific case where the fallback to ``none`` is required.

timhoffm commented:
> * edgecolor can only take one color

@story645 IIRC we said this because some aspects were represented by a Line2D, so that individual styling per violin is not supported. However, looking at the code, I only see LineCollections, which should support individual colors. Do you still remember where the Line2€ was coming in?

landoskape commented:
> Hi @landoskape, sorry didn't update last week. Rough consensus on call was:
> 
> * Add facecolor and edgecolor as parameters
> * Allow facecolor to take a list of colors
> * edgecolor can only take one color
> * Add an explicit alpha parameter and default it to the current default of 3.0

Okay, I've implemented these changes. Edgecolor can take a sequence of colors (there was discussion above to make it a sequence if possible - it is possible). 

Regarding the behavior of alpha: 
Alpha to be defaulted to 0.3. The value of alpha in the kwarg will overwrite any alpha that might have been provided in facecolor. However, if the user wants to set alpha directly in facecolor (e.g. different alphas for each facecolor in a sequence), they can set ``alpha=None`` in the kwargs, and then alpha will be determined by facecolor. 

Let me know what you all think of this.

timhoffm commented:
**TL;DR**: The API design ist mostly ok. Two requests:

- Please remove the `alpha` parameter for the reasons discussed below.
- Let's briefly consider whether `edgecolor` of `linecolor` is the more appropriate naming.

---


Sorry for letting this slip! Coming back to https://github.com/matplotlib/matplotlib/pull/27304#pullrequestreview-1728410942 and rethinking usability and consisteny: I agree that we should have some high-level color-setting keywords, and we (for now) don't need to be able to configure all individual parts (bodies, cmeans, cmins, cmaxes, cbars, cmedians, cquantiles). This means, we don't need to go the `boxplot()` way with `*props` arguments.

The question for me is what the right color-setting arguments and their scope are:

### Current situation

We have:
- the violin body as PolyCollection
- a lot of lines: the vertical line, mean, median, extrema, ...

The existing approach is to use one common color from the color cycle for everything and lighten the body through alpha=0.3.

Remark: On a high-level, it is reasonable to color all the lines the same. It's open for discussion whether we want to be able to color the lines and fill separately.

### What we could do

#### Minimal: `color`
The minimal generalization would be to introduce a `color` argument that replaces the color-cycle look. It would still take the alpha=0.3 shading for fill.

#### Separating line and fill color (this PR)
As demonstrated in this PR, we could introduce `facecolor` and `edgecolor` arguments. This is a bit of a stretch, because the facecolor applies to the PolygonCollection, whereas `edgecolor` applies to all the lines. It does not apply to the edge of the PolyCollection. But that's okish, because we cannot make the edge of the PolyCollection visible through the `violinplot` interface. While we internally use different artists, one could see a violin as semantic artists, where the edge is given through all the lines and the fill is the PolyCollection, which never has an edge.
I'm 50/50 on using `edgecolor` in that sense. The alternative would be naming it `linecolor` explicitly, which does not implicitly redefine/stretch the meaning of `egedcolor`.

I have an issue with making `alpha` available as suggested. For all artists, the `alpha` parameter is a global setting applying to both facecolor and edgecolor`. This proposed implementation does not match this. But this global parameter is also not helpful for the shading of the fill.
I propose to not make alpha configurable. Instead, if `facecolor` is not given, use the color cycle with 0.3 alpha for backward compatibility. Otherwise, use the given color as is. Not that meanwhile we support color specification as an (color, alpha) tuple, i.e. you can do `facecolor=("blue", 0.3)`, which should be good enough.

landoskape commented:
No worries, thanks for the comments! 

### Removing the alpha keyword argument
Okay, so I'll remove the alpha kwarg. We'll set the kwarg default for facecolor to be ``None``, and in this case we'll default to ``facealpha=0.3`` in the implementation rather than exposing this control in the interface. So essentially: 

```python
(..., facecolor=None, ...):
    facealpha = 0.3 if facecolor is None else None
```

### Naming: ``linecolor`` vs ``edgecolor``
I actually do think that ``linecolor`` makes more sense. It's coloring the edge of the ``PolygonCollection`` and any of the lines that show the spines / statistics, so calling it an edge feels incomplete. So I'll go ahead and call it ``linecolor`` unless you overrule. 

### Or just a minimal ``color`` argument? 
I think the debate between {``facecolor``+``linecolor``} vs {``color``} depends on what the matplotlib repo prioritizes. Adding ``linecolor`` is nice because users might want specific control (making lines black, the same color as the face, or having a specific alpha...). However, it then _requires_ users to include two kwargs. Using a minimal ``color`` interface is simpler, reduces the mental overhead of setting two parameters, but then greatly extends the boilerplate required for extra customization. 

It feels like an aesthetic choice rather than one with clear functional superiority, so I'll defer to the maintainers judgment. :)

timhoffm commented:
> So essentially:
> 
> ```python
> (..., facecolor=None, ...):
>     facealpha = 0.3 if facecolor is None else None
> ```

Yes.

> So I'll go ahead and call it `linecolor` unless you overrule.

👍 


> Or just a minimal `color` argument?

We can have all `color`, `facecolor`, `edgecolor`. That's a common pattern in the library (with defined precedence settings). I think we may end up with this here too, but I wouldn't make is part of this PR becuase the precedence handling would further complicate the review. Let's start with `facecolor` + `edgecolor` in this PR. We can have a followup that adds `color` and handles resolution of the three possible inputs to the internally used `facecolor` and `linecolor`.

landoskape commented:
How does this look?

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,108 @@ def test_violinplot_outofrange_quantiles():
+
+    # Reference image
+    ax = fig_ref.subplots(1, 5)
+    ax[0].violinplot(data, facecolor=('r', 0.5), linecolor='r', **kwargs)
```
Semantics/naming: These should be named test image, because you are testing the color parameters here. The manually tuned colors above are the reference.

landoskape replied:
yep, changed

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,108 @@ def test_violinplot_outofrange_quantiles():
+    ax = fig_test.gca()
+    ax.violinplot(data, positions=positions, facecolor=colors,
+                  linecolor=colors, **kwargs)
+    # Get all x/y axis features
+    xlim = ax.get_xlim()
+    ylim = ax.get_ylim()
+    xticks = ax.get_xticks()
+    yticks = ax.get_yticks()
+    xticklabels = ax.get_xticklabels()
+    yticklabels = ax.get_yticklabels()
+    # Ensure all x/y axis features are identical (not what this is designed to test)
+    ax.set_xlim(xlim)
+    ax.set_ylim(ylim)
+    ax.set_xticks(xticks)
+    ax.set_yticks(yticks)
+    ax.set_xticklabels(xticklabels)
+    ax.set_yticklabels(yticklabels)
```
If we have to tune all these aspects, a figure comparision may not be the right approach. Instead, you can simply read out the colors. Something like:
```
parts = violinplot(...)
assert all(mcolors.same_color(c, fc) for c, fc in zip(colors, )
...
```
Maybe even here add a helper function

```
def assert_colors_equal(colors1, colors2): 
    assert all(mcolors.same_color(c1, c2) for c1, c2 in zip(colors1, colors2)))
```
so that your assertions read a bit better `assert_colors_equal(colors, parts['cmedians'].get_facecolor())`.

landoskape replied:
smart. done

timhoffm commented:
The code is good. Left some comments on documentation and tests.

landoskape commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8635,6 +8649,17 @@ def violin(self, vpstats, positions=None, vert=None,
             'both' plots standard violins. 'low'/'high' only
             plots the side below/above the positions value.
 
+        facecolor : color or list of colors or None; see :ref:`colors_def`
+            If provided, will set the face color(s) of the violin plots.
+
+        linecolor : color or list of colors or None; see :ref:`colors_def`
+          If provided, will set the line color(s) of the violin plots (the
+          horizontal and vertical spines and body edges).
+
+            .. versionadded:: 3.11
+
+            facecolor, linecolor
```
Accepted this in the next suggestion (It's the same suggestion).

landoskape commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8717,12 +8742,38 @@ def violin(self, vpstats, positions=None, vert=None,
+            default_linecolor = cycle_color('r')
         else:
-            fillcolor = linecolor = self._get_lines.get_next_color()
+            next_color = self._get_lines.get_next_color()
+            default_facecolor = cycle_color(next_color, alpha=default_facealpha)
+            default_linecolor = cycle_color(next_color)
+
+        # Convert colors to chain (number of colors can be different from len(vpstats))
+        if facecolor is not None:
+            facecolor = cycle_color(facecolor)
+
+        if linecolor is not None:
+            linecolor = cycle_color(linecolor)
```
Agreed, not sure why it was like that. It now only uses defaults if facecolor/linecolor not provided by the user.

landoskape commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,108 @@ def test_violinplot_outofrange_quantiles():
+    # Test image
+    ax = fig_test.subplots(1, 5)
+    parts0 = ax[0].violinplot(data, **kwargs)
+    for pc in parts0['bodies']:
+        pc.set_facecolor(('r', 0.5))
+    for partname in ('cbars', 'cmins', 'cmaxes', 'cmeans', 'cmedians'):
+        if partname in parts0:
+            pc = parts0[partname]
+            pc.set_edgecolor('r')
```
done, smart

landoskape commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,108 @@ def test_violinplot_outofrange_quantiles():
+    # Ensures that setting colors in violinplot constructor works
+    # the same way as setting the color of each object manually
+    np.random.seed(19680801)
+    data = [sorted(np.random.normal(0, std, 100)) for std in range(1, 5)]
```
fair point. I made it 3 so that we can set facecolor&linecolor, just facecolor, or just linecolor

timhoffm replied:
Ok. please have a quick visual check of the generated figure to ensure the violins have still a decent width (you have 5 subplots next to each other).

landoskape replied:
It looks good with just 3 subplots (you can clearly see the details of each violin)

timhoffm commented on galleries/examples/statistics/customized_violin.py:
```diff
@@ -36,20 +36,30 @@ def set_axis_style(ax, labels):
+ax3.set_title('Customized violin plot')
+parts = ax3.violinplot(
+        data, showmeans=False, showmedians=False, showextrema=False,
+        facecolor='#D43F3A', lineolor='k')
```
```suggestion
        facecolor='#D43F3A', linecolor='k')
```

landoskape replied:
oops!

timhoffm commented on galleries/examples/statistics/customized_violin.py:
```diff
@@ -36,20 +36,30 @@ def set_axis_style(ax, labels):
 for pc in parts['bodies']:
-    pc.set_facecolor('#D43F3A')
-    pc.set_edgecolor('black')
+    pc.set_edgecolor('k')
```
```suggestion
    pc.set_edgecolor('black')
```
Let's keep the long form because it's easier to read.

landoskape replied:
done

landoskape commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,91 @@ def test_violinplot_outofrange_quantiles():
+    for part in parts_with_edgecolor:
+        colors_test = parts_test[part].get_edgecolor()
+        colors_ref = [p[part].get_edgecolor() for p in parts_ref]
+        assert_colors_equal(colors_test, colors_ref)
```
Yes, that's much more straightforward, thanks for the suggestion

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8506,6 +8507,17 @@ def violinplot(self, dataset, positions=None, vert=None,
             'both' plots standard violins. 'low'/'high' only
             plots the side below/above the positions value.
 
+        facecolor : :mpltype`color` or list of :mpltype:`color`, optional
```
```suggestion
        facecolor : :mpltype:`color` or list of :mpltype:`color`, optional
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8506,6 +8507,17 @@ def violinplot(self, dataset, positions=None, vert=None,
+
+            .. versionadded:: 3.11
+
+        linecolor : :mpltype`color` or list of :mpltype:`color`, optional
```
```suggestion
        linecolor : :mpltype:`color` or list of :mpltype:`color`, optional
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8635,6 +8649,17 @@ def violin(self, vpstats, positions=None, vert=None,
             'both' plots standard violins. 'low'/'high' only
             plots the side below/above the positions value.
 
+        facecolor : :mpltype`color` or list of :mpltype:`color`, optional
```
```suggestion
        facecolor : :mpltype:`color` or list of :mpltype:`color`, optional
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8635,6 +8649,17 @@ def violin(self, vpstats, positions=None, vert=None,
+
+            .. versionadded:: 3.11
+
+        linecolor : :mpltype`color` or list of :mpltype:`color`, optional
```
```suggestion
        linecolor : :mpltype:`color` or list of :mpltype:`color`, optional
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8748,14 +8806,15 @@ def violin(self, vpstats, positions=None, vert=None,
         bodies = []
-        for stats, pos, width in zip(vpstats, positions, widths):
+        bodies_zip = zip(vpstats, positions, widths, facecolor)
+        for stats, pos, width, facecol in bodies_zip:
```
```suggestion
        for stats, pos, width, facecolor in bodies_zip:
```
Minor, but we don't have to save letters 😄

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8748,14 +8806,15 @@ def violin(self, vpstats, positions=None, vert=None,
                             -vals + pos if side in ['both', 'low'] else pos,
                             vals + pos if side in ['both', 'high'] else pos,
-                            facecolor=fillcolor, alpha=0.3)]
+                            facecolor=facecol, edgecolor=body_edgecolor)]
```
```suggestion
                            facecolor=facecolor, edgecolor=body_edgecolor)]
```

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,83 @@ def test_violinplot_outofrange_quantiles():
+        assert all(mcolors.same_color(c1, c2)
+                   for c1, c2 in zip(colors1, colors2))
+
+    parts_with_facecolor = ["bodies"]
+    parts_with_edgecolor = ["cbars", "cmins", "cmaxes", "cmeans", "cmedians"]
+
```
```suggestion
```
Inlining is simpler because there's less state - less mental load.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,83 @@ def test_violinplot_outofrange_quantiles():
+                               linecolor=linecolors,
+                               **kwargs)
+
+    for part in parts_with_facecolor:
+        colors_test = [p.get_facecolor() for p in parts_test[part]]
+        assert_colors_equal(colors_test, mcolors.to_rgba_array(facecolors))
```
```suggestion
    body_colors = [p.get_facecolor() for p in parts_test["bodies"]]
    assert_colors_equal(body_colors, mcolors.to_rgba_array(facecolors))
```

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4021,6 +4021,83 @@ def test_violinplot_outofrange_quantiles():
+        colors_test = [p.get_facecolor() for p in parts_test[part]]
+        assert_colors_equal(colors_test, mcolors.to_rgba_array(facecolors))
+
+    for part in parts_with_edgecolor:
```
```suggestion
    for part in ["cbars", "cmins", "cmaxes", "cmeans", "cmedians"]:
```

landoskape commented:
Looks good, thanks for the cleanup! that ``facecol`` variable name was there from before I didn't have the heart to change it :)

timhoffm commented:
@landoskape Thanks for the patience to bring this through!

landoskape commented:
Woot! Thanks for working with me on it. Glad it's part of matplotlib now

timhoffm commented:
@story645 sorry that was not obvious to me. Please make a followup PR.
